src/boot_O2/loadfragment.c

```diff
@@ -1,8 +1,187 @@
+/**
+ * @file loadfragment.c
+ *
+ * Functions used to process and relocate overlays
+ *
+ * Note:
+ *  These are completly unused in favor of the functions in `loadfragment2.c`.
+ *
+ *  The main difference between them seems to be the lack of vRamEnd arguments here.
+ *  Instead they are calculated on the fly.
+ */
 #include "global.h"
 #include "system_malloc.h"
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/loadfragment/Load_Relocate.s")
+s32 gLoadLogSeverity = 2;
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/loadfragment/Load_LoadOverlay.s")
+void Load_Relocate(void* allocatedVRamAddr, OverlayRelocationSection* ovl, uintptr_t vRamStart) {
+    u32 sections[4];
+    u32* relocDataP;
+    u32 reloc;
+    uintptr_t relocatedAddress;
+    u32 i;
+    u32* luiInstRef;
+    uintptr_t allocu32 = (uintptr_t)allocatedVRamAddr;
+    u32* regValP;
+    u32* luiRefs[32];
+    u32 luiVals[32];
+    u32 isLoNeg;
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/loadfragment/Load_AllocateAndLoad.s")
+    if (gLoadLogSeverity >= 3) {}
+
+    sections[0] = 0;
+    sections[1] = allocu32;
+    sections[2] = allocu32 + ovl->textSize;
+    sections[3] = sections[2] + ovl->dataSize;
+
+    for (i = 0; i < ovl->nRelocations; i++) {
+        reloc = ovl->relocations[i];
+        relocDataP = (u32*)(sections[reloc >> 0x1E] + (reloc & 0xFFFFFF));
+
+        switch (reloc & 0x3F000000) {
+            case 0x2000000:
+                // R_MIPS_32
+                // Handles 32-bit address relocation, used for things such as jump tables and pointers in data.
+
+                // Check address is valid for relocation
+                if ((*relocDataP & 0xF000000) == 0) {
+                    *relocDataP = *relocDataP - vRamStart + allocu32;
+                } else if (gLoadLogSeverity >= 3) {
+                }
+                break;
+
+            case 0x4000000:
+                // R_MIPS_26
+                // Handles 26-bit address relocation, used for jumps and jals.
+
+                *relocDataP =
+                    (*relocDataP & 0xFC000000) |
+                    ((((((*relocDataP & 0x3FFFFFF) << 2) | 0x80000000) - vRamStart + allocu32) & 0xFFFFFFF) >> 2);
```

Comment:
These various shifts are extracting the target field from a J-type MIPS instruction (`j` and `jal`)
Not sure if we want to make that more clear or just leave it as-is.
If we had the `R4300.h` header I'd suggest changing 0x80000000 to the `K0BASE` define, but it's up to you whether you want to add that header now.

Reply:
Would it be fine to just take the header from OOT? Also would using `PHYS_TO_K0` be fine, or should I use the K0Base define directly?

Reply:
`PHYS_TO_K0` like you've done looks good. It's better to use that where possible, I failed to notice it was applicable here 😅